lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+
+	signature, err := signer.SignPROXY(jwt.PROXYSignParams{
+		SourceAddress:      sourceAddress.String(),
+		DestinationAddress: destinationAddress.String(),
+		ClusterName:        clusterName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "could not sign jwt token for PROXY line")
+	}
+
+	pl := ProxyLine{
+		Protocol:    TCP4,
+		Source:      sAddr,
+		Destination: dAddr,
+	}
+	err = pl.AddSignature([]byte(signature), signingCert)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not add signature to proxy line")
+	}
+
+	b, err := pl.Bytes()
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get bytes from proxy line")
+	}
+
+	return b, nil
+}
+
+const maxPROXYPasses = 10 // Arbitrary amount that is larger than what realistically we can get
+
+// detect finds out a type of the connection and returns wrapper that support PROXY protocol
+// We pass `casGetter` as function instead of certs themselves, so it could be called only if it's needed.
+func (m *Mux) detect(conn net.Conn) (*Conn, error) {
 	reader := bufio.NewReader(conn)
 
-	// the first attempt is to parse optional proxy
-	// protocol line that is injected by load balancers
-	// before actual protocol traffic flows.
-	// if the first attempt encounters proxy it
-	// goes to the second pass to do protocol detection
+	// Before actual protocol traffic flows, we try to parse optional PROXY protocol headers,
+	// that can be injected by load balancers or our own proxies. There can be multiple PROXY
+	// headers. After they are parsed, last pass does the actual protocol detection itself.
+	// We allow only one unsigned PROXY header from external sources, if it's enabled, and unlimited
+	// amount (within maxPROXYPasses) of signed headers from our own proxies, which take precedence.
 	var proxyLine *ProxyLine
-	for i := 0; i < 2; i++ {
+	for i := 0; i < maxPROXYPasses; i++ {
 		proto, err := detectProto(reader)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
 		switch proto {
 		case ProtoProxy:
-			if !enableProxyProtocol {
+			if !m.EnableExternalProxyProtocol {
 				return nil, trace.BadParameter("proxy protocol support is disabled")
 			}
+			// We allow only one unsigned proxy line
 			if proxyLine != nil {
 				return nil, trace.BadParameter("duplicate proxy line")
 			}
 			proxyLine, err = ReadProxyLine(reader)
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			if !enableProxyProtocol {
-				return nil, trace.BadParameter("proxy protocol support is disabled")
+			newProxyLine, err := ReadProxyLineV2(reader)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
```

Comment:
Can we hoist this out of the loop? I'm not sure we have a TTL caching layer if caching is disabled, for instance, so this could end up hitting the backend multiple times.

We should only fetch the one CA we need on demand, similarly to how it's done to authenticate connections, IMO. It's not unheard for big setups to have hundreds (or thousands) of leaf clusters, and this would fetch and unmarshal all of them on each connection.

Besides, why would we trust a leaf cluster's word on an inbound connection's source?

Reply:
I see, didn't know there are setups with that many leaf clusters 🤯 I'll look into refactoring it into getting single required CA 🤔 
And yep, definitely good to hoist it out of cycle so we guaranteed to only call it one time.

>Besides, why would we trust a leaf cluster's word on an inbound connection's source?

It's a trusted cluster, so we trust it's CA anyway

Reply:
With the way things are written here, a malicious _leaf_ cluster can produce a signature that will be trusted for the purpose of IP propagation.

A leaf cluster stores the parent's user CA (with a role map to translate role names) and not viceversa, but both parent and leaf store each other's host CAs.

Reply:
I've refactored host CA stuff - now we only try to get exact CA we know proxy cert comes from and we also check if we should trust  this cluster (basically if it's a root or local cluster).